File: core/src/main/java/com/terralite/core/registry/RegistryKey.java
package com.terralite.core.registry;

import java.util.Objects;

public record RegistryKey<T>(ResourceId id, Class<T> type) {
    public RegistryKey {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(type, "type");
    }

    public static <T> RegistryKey<T> of(ResourceId id, Class<T> type) {
        return new RegistryKey<>(id, type);
    }

    public static <T> RegistryKey<T> of(String id, Class<T> type) {
        return of(ResourceId.parse(id), type);
    }
}

File: core/src/commonMain/kotlin/com/terralite/core/registry/RegistryKey.kt
package com.terralite.core.registry

/**
 * Typed key that identifies a registry.
 *
 * The type parameter `T` provides compile-time safety when obtaining a registry
 * from [GameData]. The runtime class is intentionally omitted — the generic
 * bound alone is sufficient for correctness, and it avoids the JVM/Android
 * split between `Class<T>` and `KClass<T>`.
 */
@JvmRecord
data class RegistryKey<T>(val id: ResourceId) {

    companion object {
        @JvmStatic
        fun <T> of(id: ResourceId): RegistryKey<T> = RegistryKey(id)

        @JvmStatic
        fun <T> of(id: String): RegistryKey<T> = RegistryKey(ResourceId.parse(id))
    }
}
